modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3BlobContainer.java

```diff
@@ -317,6 +318,57 @@ public void writeBlobAtomic(OperationPurpose purpose, String blobName, BytesRefe
         writeBlob(purpose, blobName, bytes, failIfAlreadyExists);
     }
 
+    /**
+     * Perform server-side copy of a blob
+     *
+     * Server-side copy can be done for any size object, but if the object is larger than 5 GB then
+     * it must be done through a series of part copy operations rather than a single blob copy.
+     * See <a href="https://docs.aws.amazon.com/AmazonS3/latest/API/API_CopyObject.html">CopyObject</a>.
+     * @param purpose             The purpose of the operation
+     * @param sourceBlobName      The name of the blob to copy from
+     * @param targetBlobContainer The blob container to copy the blob into
+     * @param targetBlobName      The name of the blob to copy to
+     * @param failIfAlreadyExists Whether to throw a FileAlreadyExistsException if the target blob already exists
+     *                            On S3, if true, throws UnsupportedOperationException because we don't know how
+     *                            to do this atomically.
+     * @throws IOException
+     */
+    @Override
+    public void copyBlob(
+        OperationPurpose purpose,
+        String sourceBlobName,
+        BlobContainer targetBlobContainer,
```

Comment:
Should we name this destination? The feels like a more common terminology for copying.

Reply:
sure, I can rename these.

Reply:
Done in 89ae18a